### DISCUSSION TITLE
Compressor Effect / simple "Dynamic Processing" in Jamulus

### DISCUSSION THREAD

**Discussion by sunnypp** [jamulussoftware/jamulus#3323] [Ideas]
## Feature Description

Enhancing the Input Boost (1x ... 10x): What about giving the possibility to make the boost "smarter" in the sense that the "loud enough signals" won't be too amplified?

## Background / Context

This may or may not be related to #1062 , in the sense that we want the sound to be amplified. 

In my case my friends are joining Jamulus with a sub-optimal setup, as simple as a mobile phone with a wired earphone, or even just handheld it to their ears with a relatively low volume to have minimal feedback.

In such cases, they may need a Boost anywhere from **2x to 8x** so the volume gets to the green/yellow level to match with others. Such boost is needed because of the fact that their sound input device is relatively far from the sound sources (imagining playing piano with 2 hands while singing with a wired earphone hanging mid air).

And it hurts our ears when someone sneezes with 8x, or there is an accidental phone drop. This inspires a feature, that we amplify "most" signals (that we want and need), while accidental loud noises (that are over a certain limit) won't be amplified as much. _This may somehow resemble a simple Compressor audio effect._ 

Given that reverb is provided by Jamulus, and Input Boost too, so I guess it may be also appropriate to suggest providing a Compressor-like effect in Jamulus.

## Usability / How would it work?

After some thoughts, here are 2 of the possibilities that may be usuable:

### 1. Make another text field numerical input (limit or threshold for compressors) and audio signals louder than that level would be compressed.
<img width="327" alt="image" src="https://github.com/user-attachments/assets/dbb6b957-02af-4f95-821d-62386f1a2e8e">

So the existing Input Boost would behave like the "limit" is set to 0 dB and nothing is "compressed".

### 2. Make a checkbox, so when checked, a curve is used instead for calculations instead of the existing simple multiplication.
<img width="314" alt="image" src="https://github.com/user-attachments/assets/7ea4dc62-6d04-4f9e-bb77-2e82c1cb97fb">

A formula with an input boost level **n** with level of loudness **x** normalized to from 0 to 1 would be like `-(1-x)^n+1`. (so its slope at 0 would be _n_ while its slope at 1 would be 0)

## Technical implementation

It seems to me that the current input boost is handled in `CClient::ProcessAudioDataIntern`:
```
            vecsStereoSndCrd[j + 1] = static_cast<int16_t> ( iInputBoost * vecsStereoSndCrd[j + 1] );
            vecsStereoSndCrd[j]     = static_cast<int16_t> ( iInputBoost * vecsStereoSndCrd[j] );
```
As this looks cleanly decoupled / isolated, it seems I could try contribute too, though I haven't yet tried to build Jamulus on my machine.

Considering Jamulus has feedback detection, so that's why the suggestions are made intended to kept 0 dB as 0 dB to avoid breaking features that depends on 0 dB being hit on the meter.

----

What are your thoughts on the feature? Anyone also using Input Boost and suffering from unexpected loud noises that no one intended to make that happen?

Thanks for reading. And thanks for everyone here who contributes to the Jamulus community!

**Comment by ann0see** [comment]
I understand that this feature may be necessary for less ideal audio setups. (The real solution would be a better audio setup of course!)

A problem I see is that it's yet another feature which adds some complexity and must therefore be considered carefully (both from the UX and technical perspective). 

Approach 1 sounds simpler but approach 2 could also work. 
Are there audio technical advantages of one over the other one?

**Reply by sunnypp** [reply]
Thanks for the reply!

Approach 1 would have 2 or even 3 advantages when I think about it:

1. **Audio technical familiarity**: Compressors are commonly used as an audio plugin, and the idea of a threshold would sound familiar to audio engineers
2. **Audio shaping**: As long as the signals are no louder than the threshold (which is the "no accident case" for my use case), all audio signals would be simply amplified. As we look at the graph of Approach 2, the "quality" of sound would still be altered even at lower levels.
3. **Code Performance**: If we are really using "power functions", it would involve a lot of multiplications (which is done over many samples). With approach 1, each sample will be "processed" only if it crosses the threshold (saving calculations), and the calculations involved shall be a "minus", a "divide" and then "add" I guess.

If I am to think about why I recommended Approach 2 (aside from the fact that I haven't thought through it enough):
1. (Not really a sound reason) **Easier to hack**: No need to retrieve a number but just to modify the formula at that line
2. **Less intimidating**: The behaviour between enabled/disabled is clear. For Approach 1, things could go funny:
  - Someone setting threshold to -inf dB, causing everything to be over threshold and there is no input boost at all (everything is 1x). A comparison illustration is below, with the right one is "setting the threshold to be -inf dB (mistakenly)"
<img width="650" alt="image" src="https://github.com/user-attachments/assets/ca08a5df-bf5a-4c1c-a68d-784012e901bd">

  - Someone having a good setup so their signal to noise ratio **was** low. Then one tweaked input boost with threshold and end up amplifying the background noises only, since the signal (which is over the threshold by themselves without boost) are not amplified.
----
Looking forward to seeing more discussions / contributions from anyone!

**Reply by ann0see** [reply]
Ok. But approach 1 would potentially need a branch which also incurs a cost. 
I'd believe that the performance difference is negligible. 
I'd go with approach 1 probably?

**Reply by sunnypp** [reply]
> I'd go with approach 1 probably?

As a developer myself, is there anything that I could help?

----

Lemme derive the formula with approach 1, given that `MIN < THRESHOLD <= MAX`:

Let the output level be `y` and input level be `x`.
The "compressed" (flatter) line would pass through `(MAX,MAX)`, and the point that is the intersection between `y=kx` and `y=THRESHOLD`, which is `(THRESHOLD/k, THRESHOLD)` (`k` is the input boost factor)

So we could derive the slope `m` as `m = ( MAX - THRESHOLD ) / ( MAX - ( THRESHOLD / k ) )`. (∆y/∆x)

The equation (using a slope-point form formula) would then be 
```
    ( y - MAX ) / ( x - MAX ) = m
<=>                         y = m * ( x - MAX ) + MAX 
<=>                         y = MAX * ( 1 - m ) + m * x
```
which we can potentially save the first product and the value for slope before entering the loop to have negligible performance improvements on modern devices.

<img width="324" alt="image" src="https://github.com/user-attachments/assets/67ea6969-c88f-466f-ab39-b765e496fdfb">

**Reply by ann0see** [reply]
> As a developer myself, is there anything that I could help?

Sure. Check out https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md 

I'm not sure if/how much algebraic optimization bring benefit here - But precomputation seems sensible - even though the compiler may already do that? Multiplications probably take more time than addition. But it's for sure platform/architecture dependent (I don't know the cost by heart).

**Reply by sunnypp** [reply]
I have just tried on my Mac and even though my OS is on 13.1, it still builds successfully (on a re-run). (It warned me that it was built to support 10.15 only. What a pleasant surprise!)

For the algorithm part, as I am not familiar with the compiler, I don't prefer building a simple program and inspect the assembly code to check if it is optimized. Probably I will do that (negligible-)optimization in code.

For the copywriting part (e.g. what text should be shown in the Settings), what is the common practice for this codebase when implementing a new feature? Do we separate the implementation of functionality and translations into different branches or? If you are ok with it, please give me an overview of what is likely to happen, from me writing code till it's merged. 

Many thanks!

**Reply by ann0see** [reply]
You don't need to think too much about translations. This will be handled by the translators and maintainers. Just make sure to wrap text into tr().
As general rules for contribution - since we can't surely say how the process will work out for you:
1. Be aware that reviews of your PR can take a lot of time. Expect more than one month of review from multiple people and a lot of back and forth. Eventually it'll get in if everything is ok and approved by at least two maintainers. Don't be afraid to bump the PR.
3. Provide screenshots of GUI changes
4. Comment your code like the code around.
5. Open the PR quite soon once you got a decent version working to allow early review
6. Every code is GPL licensed (and see the exception for the potential iOS distribution waiver). Add yourself in the util.h file
7. Make yourself familiar with rebasing. Squash and force push changes at the end of the review with a descriptive commit message

**Comment by ann0see** [comment]
Just so you know: Please also read https://github.com/orgs/jamulussoftware/discussions/1294

Times change so we may be more or less strict on some features.

**Comment by uazure** [comment]
While having this inside jamulus would be great (as having a reverb). I think the general solution for such sort of problems is some sort of audio routing to a special software of musician's choice that does audio processing, applies effects in real time and streams it back somehow.
Such as raised in #3375 which I commented

**Reply by sunnypp** [reply]
I feel sorry for ghosting this for so long and I may not be able to come up with a PR for this...

Personally, it works well for me using REAPER with Jamulus, like passing signals through plugins to de-noise and compress and so on.

And I still have friends who'd prefer plugging in earphones to an iPhone, and immediately start some casual jamming, which caused issues like dropping the phone when being too immersed into music 😛 

I'd say that, while this is not the recommended way of jamming, some people really like how easy it is to start jamming, "as long as they have a phone with earphones". Like, muting oneself is almost the first big NO for using Jamulus, and people are still doing that so they start their Jamulus journey "feeling good". 😵 

As a side note, a friend of mine recently found that the Jamulus iOS app is gone from AppStore, and I can only hope that setting up ASIO drivers and stuffs won't be too hard for Windows users nowadays.

**Reply by ann0see** [reply]
> Jamulus iOS app is gone from AppStore

It was never an official one. And Qt6 broke the UI a lot. I can build a Testflight app in future, but don't expect it to be stable or working at all...

> and I can only hope that setting up ASIO drivers and stuffs won't be too hard for Windows users nowadays.

Unfortunately not much has changed - especially if you don't have an interface.

**Reply by ann0see** [reply]
@sunnypp I have uploaded a new build to TestFlight: https://github.com/orgs/jamulussoftware/discussions/3448
Also, there's a pending submission to the App Store. However, the review was declined due to some misunderstandings on how to use the app. I'll try to get it onto the App Store, but there's no guarantee.
